### PR TITLE
Disable "Ignoring CONTENT_IGNORE redefinition" warning

### PR DESCRIPTION
**Pull request by HybridDog**
Addresses #6276

### CODE REVIEW

**Comment by t4im** [comment]
You'll have to set its metatable to use `core.noneitemdef_default` as index, at the very least, too.

```lua
core.registered_nodes["ignore"] = setmetatable({
 	...
	name = "ignore",
	mod_origin = core.get_current_modname()
}, { __index = core.noneitemdef_default })
```

Maybe good to take a look at the c-side, too?

File: builtin/game/register.lua [hunk 1]
```diff
@@ -303,7 +303,7 @@ core.register_node(":air", {
 	groups = {not_in_creative_inventory=1},
 })
 
-core.register_node(":ignore", {
+core.registered_nodes["ignore"] = {
```
Comment (t4im) [inline builtin/game/register.lua]: Will need some fields more, if you do it without `register_node`.
```lua
 	name = "ignore",
 	mod_origin = core.get_current_modname()
```

Comment (HybridDog) [inline builtin/game/register.lua]: l don't want to copy-paste the fields, so l better do it with register_node.

Comment (HybridDog) [inline builtin/game/register.lua]: @t4im fixed

**Comment by HybridDog** [comment]
>This branch has no conflicts with the base branch
Only those with write access to this repository can merge pull requests.

Why isn't this merged?

**Comment by sofar** [comment]
OK for me, it kind of makes sense? :+1:. I can see this being useful for other things like hand, air and a few more things.

**Comment by paramat** [comment]
Build problems?

**Comment by HybridDog** [comment]
rebased

**Comment by SmallJoker** [comment]
Looks good but I would prefer another term for easier understanding. How about `is_core_content` or `is_predefined`?

**Comment by HybridDog** [comment]
l agree, l called it ```pass_to_core```.

**Comment by HybridDog** [comment]
l agree; l called it ```pass_to_core```.
#
@paramat it disappeared after rebasing

**Comment by rubenwardy** [comment]
I'm not completely sure about this. Is this because air and ignore can't be redefined?

**Comment by nerzhul** [comment]
it seems a little bit hacky for a single node, maybe we should just merge contentfeatures ?

**Comment by HybridDog** [comment]
When making mods you can use this to e.g. add the on_blast function to some node definition without redundantly passing everything to minetest core.

**Comment by paramat** [comment]
Maybe instead we can just move the warning to infostream? Or is it important to see?

**Comment by nerzhul** [comment]
No don't do that it permits to prevent node définitions conflicts for mods

**Comment by paramat** [comment]
Context: I deleted my comment, that was just before nerzhul's, where i suggested moving the message to infostream.

**Comment by HybridDog** [comment]
Could anybody donate a second approval?

**Comment by paramat** [comment]
How about just disabling the warning for the ignore node only, to keep it simple?

**Comment by HybridDog** [comment]
Do you mean explicitly avoiding printing this warning when the node is CONTENT_IGNORE?

**Comment by paramat** [comment]
Yes, but it's just a suggestion, it might not be a good idea.

**Comment by HybridDog** [comment]
l think that would be a workaround. The "ignore" node is overriden in register.lua that mods which use node definitions (e.g. minetest.registered_nodes[minetest.get_node(pos).name]) can easily treat ignore like other nodes, e.g. test whether it's walkable.
If you set minetest.registered_nodes["ignore"] to nil, ignore would be treated like unknown nodes. But doing this may add compatibility problems.

**Comment by HybridDog** [comment]
l could also temporarily override `register_item_raw` to an empty function when reregistering ignore.

**Comment by rubenwardy** [comment]
:-1: I don't like this, it feels hacky and a misfeature just to prevent this warning. Why is ignore redefined? So it appears in the tables?=

**Comment by HybridDog** [comment]
For any reason I currently do not get any warning messages (newest minetest version).
An older version (201712010933-0~ab947bd~ubuntu17.10.1) shows warnings.
The debug_log_level doesn't have an impact on what is printed in the terminal window for both versions. 
If I run bin/minetest --info, info logs are shown, error and action log messages are always shown, but the warning messages don't appear.

**Comment by paramat** [comment]
I'm still getting the message on latest master.

**Comment by HybridDog** [comment]
This is strange, when running the newest minetest from ppa I get warning messages, but when running the self built minetest they're hidden.
I just cloned minetest freshly, run cmake ., make and bin/minetest and still didn't get any warning message.
However, with the newest ppa version they're shown.

**Comment by jastevenson303** [comment]
cmake . does system wide install tho..

**Comment by HybridDog** [comment]
I think I can't reproduce it anymore…

**Comment by HybridDog** [comment]
@rubenwardy, how about now?

**Comment by paramat** [comment]
Since the 'Ignoring CONTENT_IGNORE redefinition' is harmless how about we move that to infostream? Simplest solution.

**Comment by HybridDog** [comment]
> I like rubenwardy's suggestion and will support that, please update.

l already changed it before you made this comment.

**Comment by paramat** [comment]
Rubenwardy's suggestion number 2 is not implemented in this PR, that's what is requested and what i support.

**Comment by HybridDog** [comment]
Done

**Comment by paramat** [comment]
Thanks.
@rubenwardy

**Comment by paramat** [comment]
👍

**Comment by Wuzzy2** [comment]
Hmmm.
Does this always disable the warning, or just when `ignore` is added by builtin?

IMO the warning does make sense when you try to overwrite `ignore` from a mod that is not builtin, so it should only be hidden when `builtin` tries to set it.

**Comment by rubenwardy** [comment]
That's what will happen

File: builtin/game/register.lua [hunk 2]
```diff
@@ -177,7 +179,13 @@ function core.register_item(name, itemdef)
 	--core.log("Registering item: " .. itemdef.name)
 	core.registered_items[itemdef.name] = itemdef
 	core.registered_aliases[itemdef.name] = nil
-	register_item_raw(itemdef)
+
+	-- Do not allow mods to register ignore with the engine
+	if name ~= "ignore" then
+		register_item_raw(itemdef)
+	elseif is_overriding then
+		minetest.log("warning", "Attempted redefinition of \"ignore\"")
```
Comment (red-001) [inline builtin/game/register.lua]: use the `core` namespace in builtin.

Comment (paramat) [reply builtin/game/register.lua]: Please do.

Comment (HybridDog) [reply builtin/game/register.lua]: Done, 
l also changed the comment because avoiding mods to register to core is just a redundant side effect.

**Comment by paramat** [comment]
👍

**Comment by Wuzzy2** [comment]
@rubenwardy: You answer was not hellpful. :-(
Now what will happen?

**Comment by HybridDog** [comment]
It avoids registering ignore to core if it's registered the first time in lua, i.e. in builtin.
As side effect, it gives a warning instead of attempting to register to core (which would also only give a warning, just from the core side) for subsequent ignore registrations.

**Comment by Wuzzy2** [comment]
Ah, good. That's what I wanted to hear. :-)

**Pull request merged**